x/concentrated-liquidity/internal/math/tick.go

```diff
@@ -1,52 +1,117 @@
+func TickToPrice(tickIndex, kAtPriceOne sdk.Int) (price sdk.Dec, err error) {
+	if tickIndex.IsZero() {
+		return sdk.OneDec(), nil
 	}
-	if err != nil {
-		return sdk.Dec{}, err
+
+	if kAtPriceOne.LT(types.PrecisionValueAtPriceOneMin) || kAtPriceOne.GT(types.PrecisionValueAtPriceOneMax) {
+		return sdk.Dec{}, fmt.Errorf("kAtPriceOne must be in the range (%s, %s)", types.PrecisionValueAtPriceOneMin, types.PrecisionValueAtPriceOneMax)
 	}
 
-	sqrtPrice = sqrtPriceOsmoMath.SDKDec()
+	// The formula is as follows: k_increment_distance = 9 * 10**(-k_at_price_1)
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**k_at_price_1)
+	kIncrementDistance := sdkNineDec.Mul(handleNegativeExponents(kAtPriceOne.Neg()))
+
+	// Use floor division to determine how many k increments we have passed
+	kDelta := tickIndex.ToDec().Quo(kIncrementDistance).TruncateInt()
+
+	// Calculate the current k value from the starting k value and the k delta
+	curK := kAtPriceOne.Add(kDelta)
+
+	// Knowing what our curK is, we can then figure out what power of 10 this k corresponds to
+	curIncrement := handleNegativeExponents(curK)
 
-	return sqrtPrice, nil
+	// Now, starting at the minimum tick of the current increment, we calculate how many ticks in the current k we have passed
+	numAdditiveTicks := tickIndex.ToDec().Sub(kDelta.ToDec().Mul(kIncrementDistance))
+
+	// Finally, we can calculate the price
+	price = handleNegativeExponents(kDelta).Add(numAdditiveTicks.Mul(curIncrement))
+
+	return price, nil
 }
 
 // PriceToTick takes a price and returns the corresponding tick index
-func PriceToTick(price sdk.Dec) sdk.Int {
-	tick := osmomath.BigDecFromSDKDec(price).CustomBaseLog(TickBase)
-	return tick.SDKDec().TruncateInt()
+func PriceToTick(price sdk.Dec, kAtPriceOne sdk.Int) (sdk.Int, error) {
+	if price.Equal(sdk.OneDec()) {
+		return sdk.ZeroInt(), nil
+	}
+
+	if kAtPriceOne.LT(types.PrecisionValueAtPriceOneMin) || kAtPriceOne.GT(types.PrecisionValueAtPriceOneMax) {
+		return sdk.Int{}, fmt.Errorf("kAtPriceOne must be in the range (%s, %s)", types.PrecisionValueAtPriceOneMin, types.PrecisionValueAtPriceOneMax)
+	}
+
+	// The formula is as follows: k_increment_distance = 9 * 10**(-k_at_price_1)
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**k_at_price_1)
+	kIncrementDistance := sdkNineDec.Mul(handleNegativeExponents(kAtPriceOne.Neg()))
+
+	// Initialize the total price to 1, the current k to k_at_price_1, and the number of ticks passed to 0
+	total := sdk.OneDec()
+	ticksPassed := sdk.ZeroInt()
+	currentK := kAtPriceOne
+
+	// Set the currentIncrement to the kAtPriceOne
+	curIncrement := handleNegativeExponents(currentK)
+
+	// Now, we loop through the k increments until we have passed the price
+	// Once we pass the price, we can determine what which k values we have filled in their entirety,
+	// as well as how many ticks that corresponds to
+	for total.LT(price) {
```

Comment:
q: if price is less than 1, the tick spacing is the default (as specified by kAtPriceOne), is that correct?

If so, could you expand the spec with this point, please?

Reply:
Thanks so much for brining this up, I cleaned up some incorrect negative tick (price less than 1) logic here

[5d8c1f7](https://github.com/osmosis-labs/osmosis/pull/3874/commits/5d8c1f7988fd7793638fd9d354ebde78e52ad41e)